### ISSUE TITLE
DST ROOT CA X3 expiry (30 Sep 2021) will break Composer on older PHP versions

### ISSUE THREAD

**Issue by johnstevenson** [composer/composer#10121]
Just checking to make sure that folks are aware that after Sep 30 2021 (14:01:15 UTC), PHP builds using OpenSSL versions 1.0.2 and below will be unable to connect to getcomposer.org and packagist.org (via curl or streams).

This is because these servers use the Let's Encrypt default certificate chain, which deliberately cross-signs its own ISRG Root X1 with DST ROOT CA X3.

```sh
openssl s_client -connect packagist.org:443
---
Certificate chain
 0 s:CN = packagist.org
   i:C = US, O = Let's Encrypt, CN = R3
 1 s:C = US, O = Let's Encrypt, CN = R3
   i:C = US, O = Internet Security Research Group, CN = ISRG Root X1
 2 s:C = US, O = Internet Security Research Group, CN = ISRG Root X1
   i:O = Digital Signature Trust Co., CN = DST Root CA X3
---
```

The reason is to limit breakage on older Android systems (< 7.1.1) that do not receive root certificate updates, where it transpires that the expiry date of the root CA is not actually checked.

For other clients to work with this Android-compatible chain, the self-signed ISRG Root X1 must be in the trust store and clients must recognize and use it to verify the chain. Unfortunately OpenSSL versions 1.0.2 and below do not have this capability:

* https://www.openssl.org/blog/blog/2021/09/13/LetsEncryptRootCertExpire/

## Compatibility table
This table shows which versions of Windows PHP work with the Let's Encrypt default certificate chain. The Fail/Success columns relate to the CA certificates on the client. Because Windows builds use a fixed OpenSSL version for each release (which is not updated, unlike distro builds), this is a worst case scenario.

| Windows<br>PHP version | OpenSSL version | DST_ROOT_CA_X3<br>ISRG_ROOT_X1 | ISRG_ROOT_X1 |
|-------------|----------------------|---------|---------|
| <= 5.6.10   | 1.0.1m (19 Mar 2015) | Fail    | Fail    |
| >= 5.6.11   | 1.0.1p (15 Jul 2015) | Fail    | Success |
| >= 7.2.0    | 1.1.0g (02 Nov 2017) | Success | Success |

From https://github.com/johnstevenson/dstroot which lets you test PHP connections to a server with an expired root.

## Solution
The only truly effective solution from a PHP point of view, is to create new certificates before the expiry time using the Let's Encrypt **alternate** chain, which would look like this:

```sh
openssl s_client -connect packagist.org:443
---
Certificate chain
 0 s:CN = packagist.org
   i:C = US, O = Let's Encrypt, CN = R3
 1 s:C = US, O = Let's Encrypt, CN = R3
   i:C = US, O = Internet Security Research Group, CN = ISRG Root X1
---
```
The downside is that the getcomposer.org and packagist.org web sites will break on Android < 7.1.1 (unless Firefox for Android is being used).

Additionally, composer/ca-bundle should be updated to remove DST Root CA X3 from the default bundle. This will make the bundle work for OpenSSL >=1.0.1p  (as per the above table) for servers using the Let's Encrypt default chain (and clients using the default fallback bundle).
https://github.com/composer/ca-bundle/pull/66

Also DST Root CA X3 should be removed from the installer bundles.
https://github.com/composer/windows-setup/releases/tag/v6.2.0
https://github.com/composer/getcomposer.org/pull/199
 
Alternative solutions (given enough time) could be using another certificate provider, or separating the various bits of the ecosystem that PHP connects to from the bits that are browsed.

**Comment by danepowell** [comment]
Thanks for the heads up. Is there any application-level workaround for users who are stuck on a platform with OpenSSL 1.0.2? I have not been able to find one.

Also, can anyone explain why this seems to be a problem for https://asset-packagist.org but not https://packagist.org?

Right now it seems like `composer install` works fine in this situation, but anything that reaches out to Asset Packagist (`composer update`, `composer create-project`) will fail with an error:
> curl error 60 while downloading https://asset-packagist.org/packages.json:
SSL certificate problem: certificate has expired

**Comment by Seldaek** [comment]
We re-issued our certificates yesterday before the DST Root CA X3 expired, using let's encrypt "ISRG Root X1" alternate chain (https://github.com/acmesh-official/acme.sh/wiki/Preferred-Chain).

That is why packagist.org works for you.

I unfortunately can't fix anyone else's certificates.

**Comment by Seldaek** [comment]
@danepowell you can try with `composer self-update --snapshot` though, it contains the latest version of composer/ca-bundle which removes the expired CA, it may fix the problem for you if you're using the bundled CAcert, but if not I don't think this'll do much for you.

**Comment by fritzmg** [comment]
@Seldaek just to confirm what can be done to circumvent this problem, as this is causing trouble with the Contao Manager in various hosting environments:

* Update the PHP OpenSSL extension to a sufficiently higher version (not sure which)
* Update the Root certificate of the affected service (in our case https://www.composer-resolver.cloud/) - though I have no experience what actually needs to be done here.
* Use the current snapshot of composer, which might fix the problem.

Is that accurate?

**Comment by stof** [comment]
> * Update the PHP OpenSSL extension to a sufficiently higher version (not sure what the minimum version is).

OpenSSL 1.1.0 is the version fixing the issue about multiple certificate chains where one of the reaches an expired root while the other is valid, according to blog post I saw on the subject

**Comment by johnstevenson** [comment]
@fritzmg The simplest solution is to fix the server and ensure that it supplies the Let's Encrypt **alternate** chain. This can be configured in most ACME clients (acme.sh requires ` --preferred-chain "ISRG Root X1"`, for example).

Client side fixes are using at least OpenSSL 1.1.0, OR removing the DST Root CA X3 from the trusted certificates (which allows OpenSSL down to 1.0.1p).

The current Composer snapshot contains a ca-bundle which does not include DST Root CA X3 (but this will only be used when the default CA certs location cannot be determined - on Windows for example).
